src/rpc/mining.cpp

```diff
@@ -468,21 +468,29 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
             nTransactionsUpdatedLastLP = nTransactionsUpdatedLast;
         }
 
-        // Release the wallet and main lock while waiting
+        // Release lock while waiting
```

Comment:
In commit "validation: Add missing mempool locks" (dea5dbf2d5ca0177d3a4be665790288bef40e27e)

This is hard to follow. Can you mention this change the commit description? It seems like there is no difference in behavior here except to avoid holding `g_best_block_mutex` while calling `mempool.GetTransactionsUpdated()`? Or am I misinterpreting the diff? If that is the change, it seems good to release the lock, but why is it important do to that now when it wasn't needed before?

Reply:
> It seems like there is no difference in behavior here except to avoid holding g_best_block_mutex while calling mempool.GetTransactionsUpdated()?

That is indeed the motivation and required to avoid a deadlock. (You should be able to see the deadlock stack by reverting this hunk and running the functional test suite)

Reply:
Oh, and the wallet lock was removed a long time ago, but the comment was never updated.

Reply:
This hunk as-is seems a little overcomplicated to me too if all we want to do is release `g_best_block_mutex` for the mempool operation. This patch (atop `master`) should do:
```patch
diff --git a/src/rpc/mining.cpp b/src/rpc/mining.cpp
index c0287ec17..b7f324a97 100644
--- a/src/rpc/mining.cpp
+++ b/src/rpc/mining.cpp
@@ -514,9 +514,14 @@ static UniValue getblocktemplate(const JSONRPCRequest& request)
             {
                 if (g_best_block_cv.wait_until(lock, checktxtime) == std::cv_status::timeout)
                 {
-                    // Timeout: Check transactions for update
+                    // Drop the best block mutex while accessing the mempool to avoid deadlock with
+                    // ::mempool.cs.
+                    LEAVE_CRITICAL_SECTION(lock);
+                    // Break if we've seen new transactions after the timeout has elapsed.
                     if (mempool.GetTransactionsUpdated() != nTransactionsUpdatedLastLP)
                         break;
+                    ENTER_CRITICAL_SECTION(lock);
+
                     checktxtime += std::chrono::seconds(10);
                 }
             }
```